fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/RelationalExpression.java

```diff
@@ -96,34 +94,40 @@
  * implementations of each.
  */
 @API(API.Status.EXPERIMENTAL)
-public interface RelationalExpression extends Correlated<RelationalExpression> {
+public interface RelationalExpression extends Correlated<RelationalExpression>, Typed, Narrowable<RelationalExpression> {
     @Nonnull
     static RelationalExpression fromRecordQuery(@Nonnull PlanContext context,
                                                 @Nonnull RecordQuery query) {
-        query.validate(context.getMetaData());
+        final var recordMetaData = context.getMetaData();
+        query.validate(recordMetaData);
+        final var recordTypes = context.getRecordTypes();
 
         final GroupExpressionRef<? extends RelationalExpression> baseRef;
         Quantifier.ForEach quantifier;
-        if (context.getRecordTypes().isEmpty()) {
+        if (recordTypes.isEmpty()) {
             baseRef = GroupExpressionRef.of(new FullUnorderedScanExpression(context.getMetaData().getRecordTypes().keySet()));
             quantifier = Quantifier.forEach(baseRef);
         } else {
             final var fuseRef = GroupExpressionRef.of(new FullUnorderedScanExpression(context.getMetaData().getRecordTypes().keySet()));
-            baseRef = GroupExpressionRef.of(new LogicalTypeFilterExpression(new HashSet<>(context.getRecordTypes()), Quantifier.forEach(fuseRef)));
+            baseRef = GroupExpressionRef.of(
+                    new LogicalTypeFilterExpression(
+                            new HashSet<>(recordTypes),
+                            Quantifier.forEach(fuseRef),
+                            Type.Record.fromFieldDescriptorsMap(recordMetaData.getFieldDescriptorMapFromNames(recordTypes))));
```

Comment:
Yes, I had that in mind. The fields should only merge together if they are proven to be of the same type, and should remain separately, otherwise. Maybe we could do more advanced things like recurse into records but that may give us more trouble than benefit:

```
r1 { f1 int; f2 { n1 string; n2 int } }
r2 { f0 float; f1 int; f2 { n1 string; n3 int } }
```
could become something as crazy as

```
genName { f0 float; f1 int, f2 { n1 string; n2 int, n3 int } }
```